File: soutenance-service/src/main/java/ma/enset/soutenanceservice/services/SoutenanceService.java
package ma.enset.soutenanceservice.services;

import ma.enset.soutenanceservice.entities.JuryDisponible;
import ma.enset.soutenanceservice.entities.MembreJury;
import ma.enset.soutenanceservice.entities.Soutenance;
import ma.enset.soutenanceservice.enums.RoleJury;
import ma.enset.soutenanceservice.enums.StatutSoutenance;
import org.springframework.web.multipart.MultipartFile;

import java.time.LocalDate;
import java.time.LocalTime;
import java.util.List;
import java.util.Optional;

public interface SoutenanceService {

    // ========================================================
    // CRUD
    // ========================================================

    Soutenance createSoutenance(Soutenance soutenance);

    Soutenance soumettreDemande(String titre, Long doctorantId, Long directeurId,
                                MultipartFile manuscrit, MultipartFile rapportAntiPlagiat,
                                MultipartFile autorisation);

    Soutenance updateSoutenance(Long id, Soutenance soutenance);

    void deleteSoutenance(Long id);

    Optional<Soutenance> getSoutenanceById(Long id);

    List<Soutenance> getAllSoutenances();

    List<Soutenance> getSoutenancesByDoctorant(Long doctorantId);

    List<Soutenance> getSoutenancesByDirecteur(Long directeurId);

    List<Soutenance> getSoutenancesByStatut(StatutSoutenance statut);

    // ========================================================
    // ÉTAPE 1: DIRECTEUR - Valide les prérequis
    // SOUMIS → PREREQUIS_VALIDES
    // ========================================================

    Soutenance validerPrerequisDirecteur(Long soutenanceId, String commentaire);
    Soutenance rejeterParDirecteur(Long soutenanceId, String commentaire);

    // ========================================================
    // JURYS DISPONIBLES (pour sélection dropdown)
    // ========================================================

    /**
     * Récupérer tous les jurys disponibles
     */
    List<JuryDisponible> getJurysDisponibles();

    /**
     * Récupérer les jurys disponibles par rôle
     * @param role - PRESIDENT, RAPPORTEUR, EXAMINATEUR
     */
    List<JuryDisponible> getJurysDisponiblesByRole(RoleJury role);

    /**
     * Récupérer un jury disponible par ID
     */
    Optional<JuryDisponible> getJuryDisponibleById(Long id);

    // ========================================================
    // ÉTAPE 2: DIRECTEUR - Propose le jury
    // PREREQUIS_VALIDES → JURY_PROPOSE
    // ========================================================

    Soutenance ajouterMembreJury(Long soutenanceId, MembreJury membreJury);
    Soutenance supprimerMembreJury(Long soutenanceId, Long membreId);
    Soutenance proposerJury(Long soutenanceId);

    // ========================================================
    // ÉTAPE 3: ADMIN - Valide ou refuse le jury
    // JURY_PROPOSE → AUTORISEE ou → PREREQUIS_VALIDES (retour)
    // ========================================================

    Soutenance validerJury(Long soutenanceId, String commentaire);
    Soutenance refuserJury(Long soutenanceId, String commentaire);

    // ========================================================
    // ÉTAPE 4: DIRECTEUR - Propose date de soutenance
    // ========================================================

    Soutenance proposerDateSoutenance(Long soutenanceId, LocalDate date, LocalTime heure, String lieu);

    // ========================================================
    // ÉTAPE 5: ADMIN - Planifie la soutenance
    // AUTORISEE → PLANIFIEE
    // ========================================================

    Soutenance planifierSoutenance(Long soutenanceId, LocalDate date, LocalTime heure, String lieu);
    Soutenance refuserPlanification(Long soutenanceId, String commentaire);

    // ========================================================
    // ÉTAPE 6: RÉSULTAT
    // PLANIFIEE → TERMINEE
    // ========================================================

    Soutenance enregistrerResultat(Long id, Double note, String mention, Boolean felicitations);

    // ========================================================
    // AUTRES
    // ========================================================

    Soutenance rejeterSoutenance(Long id, String motif);
    Soutenance soumettreRapportRapporteur(Long soutenanceId, Long membreJuryId, Boolean avisFavorable, String commentaire);

    // Legacy
    Soutenance verifierPrerequisEtSoumettre(Long id);
    Soutenance autoriserSoutenance(Long id, String commentaire);
}

File: portail-doctorat-frontend/src/app/core/services/soutenance.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';
import { Soutenance } from '@core/models/soutenance.model';

@Injectable({
    providedIn: 'root'
})
export class SoutenanceService {
    private baseUrl = `${environment.soutenanceServiceUrl}/soutenances`;

    constructor(private http: HttpClient) {
        console.log('🔧 SoutenanceService URL:', this.baseUrl);
    }

    // ========== CRUD ==========

    getAllSoutenances(): Observable<Soutenance[]> {
        return this.http.get<Soutenance[]>(this.baseUrl);
    }

    getSoutenanceById(id: number): Observable<Soutenance> {
        return this.http.get<Soutenance>(`${this.baseUrl}/${id}`);
    }

    getSoutenancesByDoctorant(doctorantId: number): Observable<Soutenance[]> {
        return this.http.get<Soutenance[]>(`${this.baseUrl}/doctorant/${doctorantId}`);
    }

    getSoutenancesByDirecteur(directeurId: number): Observable<Soutenance[]> {
        return this.http.get<Soutenance[]>(`${this.baseUrl}/directeur/${directeurId}`);
    }

    createSoutenance(soutenance: Partial<Soutenance>): Observable<Soutenance> {
        return this.http.post<Soutenance>(this.baseUrl, soutenance);
    }

    soumettreDemande(formData: FormData): Observable<Soutenance> {
        return this.http.post<Soutenance>(`${this.baseUrl}/soumettre`, formData);
    }

    // ========== ÉTAPE 1: DIRECTEUR valide prérequis (SOUMIS → PREREQUIS_VALIDES) ==========

    validerPrerequisDirecteur(id: number, commentaire: string): Observable<Soutenance> {
        return this.http.put<Soutenance>(`${this.baseUrl}/${id}/valider-prerequis`, { commentaire });
    }

    rejeterDemandeDirecteur(id: number, commentaire: string): Observable<Soutenance> {
        return this.http.put<Soutenance>(`${this.baseUrl}/${id}/rejeter-directeur`, { commentaire });
    }

    // ========== ÉTAPE 2: ADMIN autorise la demande (PREREQUIS_VALIDES → AUTORISEE) ==========

    autoriserSoutenance(id: number, commentaire?: string): Observable<Soutenance> {
        return this.http.put<Soutenance>(`${this.baseUrl}/${id}/autoriser`, { commentaire });
    }

    // ========== ÉTAPE 3: DIRECTEUR propose le jury (AUTORISEE → JURY_PROPOSE) ==========

    getMembresJuryByRole(role: string): Observable<any[]> {
        return this.http.get<any[]>(`${this.baseUrl}/jury/disponibles/${role}`);
    }

    getJurysDisponibles(): Observable<any[]> {
        return this.http.get<any[]>(`${this.baseUrl}/jury/disponibles`);
    }

    ajouterMembreJury(soutenanceId: number, membre: any): Observable<Soutenance> {
        return this.http.post<Soutenance>(`${this.baseUrl}/${soutenanceId}/jury`, membre);
    }

    supprimerMembreJury(soutenanceId: number, membreId: number): Observable<Soutenance> {
        return this.http.delete<Soutenance>(`${this.baseUrl}/${soutenanceId}/jury/${membreId}`);
    }

    proposerJury(id: number): Observable<Soutenance> {
        return this.http.put<Soutenance>(`${this.baseUrl}/${id}/proposer-jury`, {});
    }

    // ========== ÉTAPE 4: ADMIN valide le jury (JURY_PROPOSE) ==========

    validerJury(id: number, commentaire: string): Observable<Soutenance> {
        return this.http.put<Soutenance>(`${this.baseUrl}/${id}/valider-jury`, { commentaire });
    }

    refuserJury(id: number, commentaire: string): Observable<Soutenance> {
        return this.http.put<Soutenance>(`${this.baseUrl}/${id}/refuser-jury`, { commentaire });
    }

    // ========== ÉTAPE 5: ADMIN planifie la soutenance (→ PLANIFIEE) ==========

    planifierSoutenance(id: number, data: { dateSoutenance: string; heureSoutenance?: string; lieuSoutenance?: string }): Observable<Soutenance> {
        return this.http.put<Soutenance>(`${this.baseUrl}/${id}/planifier`, data);
    }

    // ========== ÉTAPE 6: ADMIN enregistre le résultat (PLANIFIEE → TERMINEE) ==========

    enregistrerResultat(id: number, data: { mention: string; felicitations?: boolean }): Observable<Soutenance> {
        return this.http.put<Soutenance>(`${this.baseUrl}/${id}/resultat`, data);
    }

    // ========== AUTRES ==========

    rejeterSoutenance(id: number, motif: string): Observable<Soutenance> {
        return this.http.put<Soutenance>(`${this.baseUrl}/${id}/rejeter`, { motif });
    }

    getDocumentUrl(filePath: string): string {
        if (!filePath) return '';

        let filename = filePath;
        if (filePath.includes('uploads/soutenances/')) {
            filename = filePath.split('uploads/soutenances/').pop() || filePath;
        } else if (filePath.includes('/')) {
            filename = filePath.split('/').pop() || filePath;
        }

        // ✅ URL correcte : /api/soutenances/files/{filename}
        return `${environment.soutenanceServiceUrl}/soutenances/files/${filename}`;
    }

    openDocument(path: string): void {
        if (path) {
            window.open(this.getDocumentUrl(path), '_blank');
        }
    }
}